design/lakectl-testing.md

```diff
@@ -0,0 +1,102 @@
+# lakectl Testing
+
+## Purpose
+
+This document describes the design of a testing framework aimed at testing our CLI, namely `lakectl`.
+It specifies the requirements from the system and a set of technical questions we need to answer. Each question (in fact, each option), is accompanied by pros and cons, to help us understand which options will answer our needs the best. These are obviously not the only aspects to these decisions, and so, this document and the decisions derived from, are subject to reviews and suggestions.
+This document does not specify the cli tests to run, but the system intended to run them
+
+## Requirements
+* Usable both locally and as part of CI/CD
+* Ease of use - adding new tests should be as easy as possible. Moreover, adding new tests, or modifying existing ones, should not draw modifications to the testing framework itself
+  * It is obvious that changing the test **format** (e.g. expanding) will require framework modifications, in order to support the modified format
+* Decoupling of Framework and Data - the test data format should contain all the required info for the test itself, regardless of the framework design. In other words, we would like the ability to rewrite/replace the framework and still use the same tests data
+*
+
+## What We Want To Test 
+
+The goal is to the test `lakectl` correctness. At the very least we would like to verify that:
+* CLI input is parsed correctly
+* The correct requests are sent to the server (if any)
+* Server result is formatted correctly
+* Both erroneous input and error results from the server are formatted correctly
+* CLI Internal errors are handled correctly
+
+This testing framework is not designed to perform end-to-end tests (i.e. server correctness) or performance, for that matter
+
+## Architecture
+
+### Architecture Outline
+
+![lakectl testing architecture](./diagrams/lakectl-testing-arch.png)
+
+### Test Case Structure
+
+The test case structure should contain at least the following details:
+* pre-condition: the system state right before the tested command starts. This can be a function name to call, a mock type to use or a file containing lakectl commands to run in order to bring the system to that state (in case real server is in use)
+* lakectl command: a string with the `lakectl` command to test.
```

Comment:
Commented by @itaiad200 
Must be just a single command? What's the downside of writing a more elaborated test that covers more than a single command?

Reply:
That is absolutely correct, when using a live server. When using mocks, you can easily generate whatever "`lakefs`" output that you want, and don't really need the dependency between the two commands.
I can totally see where are the difficulties in my aproach (as you wrote - it treats the tests more like unti tests and less like system tests) and the true benefits of letting the real `lakefs` do its thing.
Moreover, I did wrote some simillar mocks in the past, where the mock itself verified the input it received (the client request) and returned hardcoded, tailored output which the client processed and generated its own, "real", output, which was finally verified by the test running tool. I must admit, though, that the testing framework was not as generic as one may hope, and it forced us to write and rewrite the mocks very often, so I would definitley don't want to go that way!